vyper/venom/analysis/loop_detection.py

```diff
@@ -0,0 +1,45 @@
+from vyper.venom.analysis.analysis import IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.basicblock import IRBasicBlock
+
+
+class LoopDetection(IRAnalysis):
+    """
+    Detects loops and computes basic blocks sets
+    which comprised these loops
+    """
+
+    # key = start of the loop (last bb not in the loop), value all the block that loop contains
+    loops: dict[IRBasicBlock, list[IRBasicBlock]]
+
+    def analyze(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.loops: dict[IRBasicBlock, list[IRBasicBlock]] = dict()
+        done: dict[IRBasicBlock, bool] = dict(
+            (bb, False) for bb in self.function.get_basic_blocks()
+        )
+        entry = self.function.entry
+        self.dfs(entry, done, [])
+
+    def invalidate(self):
+        return super().invalidate()
+
+    def dfs(self, bb: IRBasicBlock, done: dict[IRBasicBlock, bool], path: list[IRBasicBlock]):
```

Comment:
Couldn't this `done` be promoted to a class member?

Reply:
I think it could. Should it be promoted even though it is only used for `dfs` method and is reset before every run of `analyze` and is not necessary for consumer of this analysis?